File: solutions/java/src/main/java/playground/algorithm/CombinationSum.java
package playground.algorithm;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class CombinationSum {

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);

        //原地去重并返回数组的长度
        int length = 1;
        for (int i = 0; i < candidates.length; i++) {
            if (candidates[i] != candidates[length - 1]) {
                candidates[length++] = candidates[i];
            }
        }

        List<List<Integer>> result = new ArrayList<>();
        combinationSum(candidates, 0, length, target, new ArrayList<>(), result);
        return result;
    }

    private void combinationSum(int[] candidates, int start, int length, int target, List<Integer> combination, List<List<Integer>> result) {
        if (target < 0) {
            return;
        }
        if (target == 0) {
            result.add(new ArrayList<>(combination));
        }

        for (int i = start; i < length; i++) {
            combination.add(candidates[i]);
            combinationSum(candidates, i, length, target - candidates[i], combination, result);
            combination.remove(combination.size() - 1);
        }
    }
}

File: solutions/go/src/playground/algorithm/combination_sum.go
package algorithm

import "sort"

func combinationSum(candidates []int, target int) [][]int {
	sort.Ints(candidates)

	length := 1
	for _, n := range candidates {
		if n != candidates[length-1] {
			candidates[length] = n
			length++
		}
	}

	result := make([][]int, 0)
	combinationSumR(candidates[0:length], 0, make([]int, 0), target, &result)
	return result
}

func combinationSumR(candidates []int, start int, combination []int, target int, result *[][]int) {
	if target < 0 {
		return
	}
	if target == 0 {
		tmp := make([]int, len(combination))
		copy(tmp, combination)
		*result = append(*result, tmp)
		return
	}

	for i := start; i < len(candidates); i++ {
		combinationSumR(candidates, i, append(combination, candidates[i]), target-candidates[i], result)
	}
}
